R/irace_helpers.R

```diff
@@ -0,0 +1,64 @@
+paradoxToIrace = function(ps){
+  assertClass(ps, "ParamSet")
+  # what about ParamUty = vector numeric/real
+  class_lookup = data.frame(paradox = c("ParamLgl","ParamInt","ParamDbl","ParamFct"),
+                            irace = c("c","i","r","c"), stringsAsFactors = FALSE)
+  name <- ps$ids()
+  type <- class_lookup[match(ps$class, class_lookup$paradox), 2]
+  range <- getIraceRange(ps)
+  if(ps$has_deps){
+    condition <- getIraceCondition(ps)
+  } else {
+    condition = NULL
+  }
+
+  # does mlr3tuning deal with trafo internally or passed to irace? irace only handle log
+  par.tab <- paste(name, '""', type, range, condition, collapse = "\n")
+  irace::readParameters(text = par.tab)
+}
+getIraceRange = function(ps){
+  rng = data.table(lower = ps$lower, upper = ps$upper, lvl = ps$levels)
+
+  apply(rng, 1, function(x){
+    if(is.na(x[[1]]))
+      return(paste0("(",paste0(x[[3]], collapse = ","),")"))
+    else
+      return(paste0("(",x[[1]],",",x[[2]],")"))
+  })
+}
+getIraceCondition = function(ps){
+  cond = apply(ps$deps, 1, function(x){
+    if(class(x[[3]])[1] == "CondEqual"){
+      condition = paste("|",x[[2]], "==", x[[3]]$rhs)
+    } else {
+      condition = paste("|",x[[2]],"%in%", paste0("c(",paste0(x[[3]]$rhs, collapse = ","),")"))
+    }
+  })
+
+  condition = data.table(id = ps$ids(), cond = "")
+  ind = match(ps$deps$id, condition$id)
+  condition$cond[ind] = cond
+  condition$cond
+}
+
+makeScenario = function(instance){
+  list(
+    targetRunner = targetRunner,
+    logFile = tempfile(),
+    instances = list(instance),
```

Comment:
Now my other comments were nitpicking, but I believe this is a serious error. The `instances` of irace are "challenges" to a given parameter configuration that this configurations is evaluated on. The results of these challenges are compared between different parameter configurations (in a pairwise test afair).

`instances` are not what we call a `TuningInstance`. Instead, they should be different resampling instances (or other things that are drawn at random from a population of which we are trying to get a configuration with best average performance: could be different tasks). I guess our `TuningInstance` the way we have it implemented currently doesn't work like that because it does not accept resampling instances as parameters. `TuningInstance` will probably have to be adjusted.

Reply:
So this is exactly the part that I needed checking. The thing that threw me is actually the [mlr2 implementation]. You might be able to interpret this better than me but:

1. Within the individual experiment we have the [first instance always being selected](https://github.com/mlr-org/mlr/blob/1cb54c7e3750f2a20afc0740f368e64a3a5df76e/R/tuneIrace.R#L14-L15).
2. Then [new resample instances being created](https://github.com/mlr-org/mlr/blob/1cb54c7e3750f2a20afc0740f368e64a3a5df76e/R/tuneIrace.R#L29) up to the number specified in `n.instances`

I'm not familiar enough with the mlr2 or mlr3 implementations to know how creating multiple resamples from an instance in mlr2 differs from what is done in mlr3tuning. What we could do is replicate the `TuningInstance` in by `n.instances` if this creates different resampling schemes, or challenges as you put it.